From f65a46c9e5503c1c769ddbbb087d6c0258b63a2c Mon Sep 17 00:00:00 2001
From: Murala S Swamy <ssmurala@habana.ai>
Date: Wed, 28 Aug 2024 13:21:53 +0300
Subject: [PATCH] [FSW-8694] fs1: add generic api to verify hash

Generic API which accepts manifest index and
firmware image type as inputs, it extracts
img desc from manifest and provide required
inputs for key hash library api.
This API can be used to verify key hash of any
firmware image which has img desc inside
manifest.

Change-Id: Idc7ffd018c8671d32b0aaa73c7ed3d603582fa92
---

diff --git a/common/secure_boot.c b/common/secure_boot.c
index 395d797..9f00e53 100644
--- a/common/secure_boot.c
+++ b/common/secure_boot.c
@@ -12,6 +12,8 @@
 #include <boot_rom.h>
 #include <soc.h>
 #include <dev_security.h>
+#include <boot_desc.h>
+#include <common.h>
 
 int secure_boot_verify_image_hash384(uint8_t *exp_hash, unsigned char *img, uint32_t img_len)
 {
@@ -66,3 +68,28 @@
 
 	return SECUREBOOT_VERIFY_FAILED;
 }
+
+bool boot_rom_verify_fw_image_hash384(enum hl_fw_manifest_type manifest_idx, enum hl_fw_img_type img_type)
+{
+	uint32_t act_img_size, loadaddr_arc;
+	unsigned char *image;
+	struct boot_desc desc = {0};
+
+	/* get image descriptor from the manifest */
+	if (boot_desc_get_img_desc(manifest_idx, img_type, &desc, 0) != BOOT_ROM_SUCCESS) {
+		return false;
+	}
+
+	act_img_size = desc.img_size;
+	act_img_size = act_img_size - CRC32_SIZE;
+
+	/* TODO: Instead of loadaddr_arc we will use dma dest addr */
+	loadaddr_arc = (uint32_t)boot_rom_get_arc_addr(desc.loadaddr);
+	image = (unsigned char *)(loadaddr_arc);
+
+	if (secure_boot_verify_image_hash384(desc.chksum, image, act_img_size) != 0) {
+		return false;
+	}
+
+	return true;
+}
diff --git a/common/secure_boot.h b/common/secure_boot.h
index b336706..29010cb 100644
--- a/common/secure_boot.h
+++ b/common/secure_boot.h
@@ -16,5 +16,7 @@
 int secure_boot_verify_image_signature(const uint8_t *image, uint32_t image_len, uint8_t *key,
 				uint32_t key_len, uint8_t *sig, uint32_t sig_len,
 				uint8_t sec_ver, uint8_t key_index, uint8_t rma_manifest);
+bool boot_rom_verify_fw_image_hash384(enum hl_fw_manifest_type manifest_idx,
+	enum hl_fw_img_type img_type);
 
 #endif /* _SECURE_BOOT_H_ */
